fix(ci): enable module downloads in CIFuzz build

puwun opened the pull request:
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] Thanks for contributing!
- [x] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)

---

<!-- Description of change -->

The scheduled CIFuzz runs have been failing since the end of September. The December build logs show:
```
cannot find module providing package github.com/AdamKorcz/go-118-fuzz-build/testing: import lookup disabled by -mod=vendor
(Go version in go.mod is at least 1.14 and vendor directory exists.)
```
The `go-118-fuzz-build/testing` package is required by `compile_native_go_fuzzer` but is not present in the vendor directory. When Go runs with `-mod=vendor` (the default when a vendor directory exists), it cannot download missing dependencies from the module proxy.

This change exports `GOFLAGS="-mod=mod"` before running `compile_native_go_fuzzer`, allowing Go to download the required package from the module proxy.

---
Fixes: #43477 

<!-- Release Notes -->
```release-note
ci: Fix CIFuzz build by enabling Go module downloads for dependencies not in vendor directory
```

joestringer commented:
Thanks for the patch!

Sorry, there's a bit of bitrot going on in that harness at the moment. Here's the latest failure after applying the change in this PR:

https://github.com/cilium/cilium/actions/runs/20529083512/job/59433731578?pr=43504#step:4:618

```
pkg/policy/l4_filter_test_fuzz.go:487:15: undefined: OriginForTest
```

It looks like https://github.com/cilium/cilium/commit/0399e2f47d951 didn't take into account that this function was being used by fuzzers outside the package(?). Perhaps we can work around that or just revert that commit.

For broader context also, the most recent attempt to fix up CIFuzz was over here: https://github.com/cilium/cilium/pull/41780 . I haven't had a chance to really sit down and try to dig through and figure out how to get these tests back up. Previously I've had a lot more success with `make fuzz` than with these external harnesses (though I just noticed that was broken, so I've fixed it up here: https://github.com/cilium/cilium/pull/43555).

puwun commented:
@joestringer the `OriginForTest` function is defined in `pkg/policy/origin_test.go` but oss-fuzz build uses go build (not go test), which doesn't include `_test.go` files. The existing `oss-fuzz-build.sh` script already handles this by creating symlinks from `*_test.go ` to `*_fuzz.go` for other policy test files, but `origin_test.go` was not included there

**Proposed fix:** 
add the following symlink to `test/fuzzing/oss-fuzz-build.sh`:
```
ln -s "$SRC"/cilium/pkg/policy/origin_test{,_fuzz}.go
```
this follows the same pattern already used for other test files in the script:
```
ln -s "$SRC"/cilium/pkg/policy/distillery_test{,_fuzz}.go
ln -s "$SRC"/cilium/pkg/policy/l4_filter_test{,_fuzz}.go
# ... and so on
```

lmk if any changes to this, otherwise i'll raise a PR shortly

joestringer commented:
Sounds good to me :+1:

joestringer commented:
Other thing I will suggest is to rebase on top of the latest upstream branch and force push your branch to get rid of the merge commits. We don't integrate merge commits into Cilium. Note that it's not required to be up to date against the base branch, we receive a high rate of submissions in Cilium so it is not realistic to always have your PR be based on the tip of the `main` branch.

puwun committed:
fix(ci): enable module downloads and add origin_test symlink for CIFuzz

The CIFuzz builds were failing due to:
1. go-118-fuzz-build/testing package not in vendor directory
2. OriginForTest function not available during fuzz builds

This fix:
- Adds GOFLAGS=-mod=mod to allow downloading missing dependencies
- Adds symlink for origin_test.go to make OriginForTest available

Fixes: #43477
Signed-off-by: Pavan More <pavansmore05@gmail.com>

puwun commented:
@joestringer I've pushed the changes, please take a look whenever you get a chance

puwun commented:
yes, I've added symlinks for both `l4_filter_deny_test.go` and `repository_deny_test.go` (also uses `denyPerSelectorPolicy`)

joestringer commented:
The issue above seems related to the following item: https://github.com/AdamKorcz/go-118-fuzz-build/issues/28 .

I wonder whether it's worthwhile to merge this change into the tree first since it's a step in the right direction, then pursue moving to `compile_native_go_fuzzer_v2` as proposed in https://github.com/cilium/cilium/pull/41780 ? Or alternatively you might consider cherry-picking the change from that PR into here and iterating on top to address the error reports from that PR.

(One benefit of merging the PR as-is is that it should allow a quicker iteration cycle. Currently as a brand new contributor, your PRs require manual approval to run CI each time you push a fresh version)
